### PR TITLE
[core] Fix using `in` expression with `filter`

### LINKED ISSUE #16262: Unable to use in expression as layer filter
The `in` expression operator added in #16162 can only be used with paint and layout properties, not with layer filters. To reproduce, load this style and observe that nothing is visible:

<details>
<summary>colors.json</summary>

```json
{
  "version": 8,
  "zoom": 0,
  "center": [0, 0],
  "sources": {
    "quadrants": {
      "type": "geojson",
      "data": {
        "type": "FeatureCollection",
        "features": [
          {
            "type": "Feature",
            "properties": {
              "name": "ABC",
              "color": "red"
            },
            "geometry": {
              "type": "Polygon",
              "coordinates": [[
                [-180, -90],
                [-180, 0],
                [0, 0],
                [0, -90],
                [-180, -90]
              ]]
            }
          },
          {
            "type": "Feature",
            "properties": {
              "name": "DEF",
              "color": "#00ff00"
            },
            "geometry": {
              "type": "Polygon",
              "coordinates": [[
                [-180, 90],
                [-180, 0],
                [0, 0],
                [0, 90],
                [-180, 90]
              ]]
            }
          },
          {
            "type": "Feature",
            "properties": {
              "name": "GHI",
              "color": "#0000ff"
            },
            "geometry": {
              "type": "Polygon",
              "coordinates": [[
                [180, -90],
                [180, 0],
                [0, 0],
                [0, -90],
                [180, -90]
              ]]
            }
          },
          {
            "type": "Feature",
            "properties": {
              "name": "GHI",
              "color": "#ffff00"
            },
            "geometry": {
              "type": "Polygon",
              "coordinates": [[
                [180, 90],
                [180, 0],
                [0, 0],
                [0, 90],
                [180, 90]
              ]]
            }
          }
        ]
      }
    }
  },
  "glyphs": "local://glyphs/{fontstack}/{range}.pbf",
  "layers": [
    {
      "id": "colors",
      "type": "fill",
      "source": "quadrants",
      "filter": ["in", "#", ["get", "color"]],
      "paint": {
        "fill-color": ["get", "color"]
      }
    }
  ]
}
```

</details>

GL JS v1.8.1 | Expected | GL Native v1.3.0
----|----|----
gl-js | expected | actual

`mbgl::style::conversion::isExpression()` always returns false for the filter `["in", "#", ["get", "color"]]`, because `in` is also a [legacy filter operator](https://docs.mapbox.com/mapbox-gl-js/style-spec/other/#set-membership-filters):

https://github.com/mapbox/mapbox-gl-native/blob/305df25420e8ec592869d1048ee05a9fb2215bdd/src/mbgl/style/conversion/filter.cpp#L56-L57

/ref https://github.com/mapbox/mapbox-gl-native-ios/pull/183#issuecomment-593731631
/cc @jmkiley @Chaoba @chloekraw

### PR DESCRIPTION
**Pull request by Chaoba**
Resolves #16262 
`mbgl::style::conversion::isExpression()` always returns false for expression `in`, this PR fix this issue.

The test in added in android test app in this pr: https://github.com/mapbox/mapbox-gl-native-android/pull/218

I'm not sure where should add test cases for this change, @alexshalamov any suggestions for it?

### CODE REVIEW

File: src/mbgl/style/conversion/filter.cpp [hunk 1]
```diff
@@ -53,9 +53,14 @@ bool isExpression(const Convertible& filter) {
...
         return operand && *operand != "$id" && *operand != "$type";
 
-    } else if (*op == "in" || *op == "!in" || *op == "!has" || *op == "none") {
+    } else if (*op == "!in" || *op == "!has" || *op == "none") {
         return false;
 
+    } else if (*op == "in") {
+        optional<std::string> str = toString(arrayMember(filter, 1));
+
+        return arrayLength(filter) >= 3 && (!str || isArray(arrayMember(filter, 2)));
```
Comment (1ec5) [inline src/mbgl/style/conversion/filter.cpp]: While you’re at it, can you incorporate the change in mapbox/mapbox-gl-js#9374 too?

**Comment by 1ec5** [comment]
> I'm not sure where should add test cases for this change

mapbox/mapbox-gl-js#9374 adds relevant test cases to the style specification unit tests. Would it be feasible to update the submodule to include that change?

File: src/mbgl/style/conversion/filter.cpp (hunk 1 above)
Comment (Chaoba) [reply src/mbgl/style/conversion/filter.cpp]: I did incorporate that change.

**Comment by alexshalamov** [comment]
@Chaoba PR looks good, could you add render test for that in gl-js? Something similar to an example in https://github.com/mapbox/mapbox-gl-native/issues/16262

File: src/mbgl/style/conversion/filter.cpp (hunk 1 above)
Comment (1ec5) [reply src/mbgl/style/conversion/filter.cpp]: Ah, my bad, I overlooked that part because of the variable name `str`, which I misunderstood at first.

**Comment by 1ec5** [comment]
Unfortunately, rendering tests are now failing due to updating the GL JS submodule, which brought in a test case for mapbox/mapbox-gl-js#9198 along with the test case for mapbox/mapbox-gl-js#9374 that we wanted.

**Comment by alexshalamov** [comment]
> Unfortunately, rendering tests are now failing due to updating the GL JS submodule, which brought in a test case for mapbox/mapbox-gl-js#9198

@Chaoba could you please add failing tests related to  mapbox/mapbox-gl-js#9198 `test/integration/query-tests/evaluated/*` to ignore list https://github.com/mapbox/mapbox-gl-native/blob/master/metrics/ignores/platform-all.json

**Comment by Chaoba** [comment]
> @Chaoba PR looks good, could you add render test for that in gl-js? Something similar to an example in #16262

@alexshalamov Created a [ticket](https://github.com/mapbox/mapbox-gl-js/issues/9386) in gl-js to track it.

**Comment by Chaoba** [comment]
@alexshalamov I didn't change these files, why sanity-checks failed?

**Comment by Chaoba** [comment]
The issue in this pr is that:
1. If we update gl-js, the nitpick will fail.
2. If we apply  nitpick patch, then the test case will fail
```
2: [ RUN      ] Annotations.DebugSparse
2: ../test/src/mbgl/test/util.cpp:49: Failure
2: Expected: (pixels / (expected.size.width * expected.size.height)) <= (imageThreshold), actual: 0.015564 vs 0.0002
2: [  FAILED  ] Annotations.DebugSparse (103 ms)
```
@tmpsantos  What should we do to let this pr merged?

**Comment by 1ec5** [comment]
The failing test came in through mapbox/mapbox-gl-js#9397 as a result of #16293, which has landed. It should pass once you rebase this branch onto master.

**Comment by chloekraw** [comment]
@tmpsantos CI is passing, good to merge?

**Comment by chloekraw** [comment]
Ack, forgot the changelog. @Chaoba would you mind adding an entry in a separate pr?

**Comment by pozdnyakov** [comment]
Change log entry proposal
```
### Bugs

Fixed using of the `in` expression as a layer filter ([#16272](https://github.com/mapbox/mapbox-gl-native/pull/16272))

The bug was caused by `mbgl::style::conversion::isExpression()` always returning `false` for the `in` expression.
```

**Pull request merged**